langchain/src/experimental/masking/parser.ts

```diff
@@ -0,0 +1,139 @@
+   */
+  async parse(message: string): Promise<string> {
+    this.config.onMaskingStart?.(message);
+
+    // Check if there are any transformers added to the parser. If not, throw an error
+    // as masking requires at least one transformer to apply its logic.
+    if (this.transformers.length === 0) {
+      throw new Error(
+        "MaskingParser.parse Error: No transformers have been added. Please add at least one transformer before parsing."
+      );
+    }
+
+    if (typeof message !== "string") {
+      throw new TypeError(
+        "MaskingParser.parse Error: The 'message' argument must be a string."
+      );
+    }
+
+    // Initialize the variable to hold the progressively masked message.
+    // It starts as the original message and gets transformed by each transformer.
+    let processedMessage = message;
+
+    // Iterate through each transformer added to the parser.
+    this.transformers.forEach((transformer) => {
+      // Apply the transformer's transform method to the current state of the message.
+      // The transform method returns a tuple containing the updated message and state.
+      // The state is a map that tracks the original values of masked content.
+      // This state is essential for the rehydration process to restore the original message.
+      [processedMessage, this.state] = transformer.transform(
+        processedMessage,
+        this.state
+      );
+    });
+
+    this.config.onMaskingEnd?.(processedMessage);
+    // Return the fully masked message after all transformers have been applied.
+    return processedMessage;
+  }
+
+  /**
+   * Rehydrates a masked message back to its original form.
+   * This method sequentially applies the rehydration logic of each added transformer in reverse order.
+   * It relies on the state map to correctly map the masked values back to their original values.
+   *
+   * The rehydration process is essential for restoring the original content of a message
+   * that has been transformed (masked) by the transformers. This process is the inverse of the masking process.
+   *
+   * @param message - The masked message to be rehydrated.
+   * @returns The original (rehydrated) version of the message.
+   */
+  async rehydrate(
+    message: string,
+    state?: Map<string, string>
+  ): Promise<string> {
+    this.config.onRehydratingStart?.(message);
+
+    if (typeof message !== "string") {
+      throw new TypeError(
+        "MaskingParser.rehydrate Error: The 'message' argument must be a string."
+      );
+    }
+    // Check if any transformers have been added to the parser.
+    // If no transformers are present, throw an error as rehydration requires at least one transformer.
+    if (this.transformers.length === 0) {
+      throw new Error(
+        "MaskingParser.rehydrate Error: No transformers have been added. Please add at least one transformer before rehydrating."
+      );
+    }
+
+    if (state && !(state instanceof Map)) {
```

Comment:
Our linter does not allow `instanceof` since bundlers/chunking can make it unreliable in certain cases - I think it's fine here but you'll need to disable it with a comment:

https://github.com/langchain-ai/langchainjs/blob/dfd32715c44e9bc63f98bf5b8fd93dbd62b3320c/langchain-core/src/prompts/chat.ts#L502